modules/web/client/src/main/scala/observe/ui/components/MainApp.scala

```diff
@@ -0,0 +1,212 @@
+
+object MainApp:
+  private val ConfigFile: Uri = uri"/environments.conf.json"
+  private val ApiBaseUri: Uri = uri"/api/observe"
+  private val EventWsUri: Uri =
+    Uri.unsafeFromString("wss://" + dom.window.location.host + ApiBaseUri + "/events")
+
+  // Set up logging
+  private def setupLogger(level: LogLevelDesc): IO[Logger[IO]] = IO:
+    LogLevelLogger.setLevel(level)
+    LogLevelLogger.createForRoot[IO]
+
+  // Define React routing
+  private val (router, routerCtl) =
+    RouterWithProps.componentAndCtl(BaseUrl.fromWindowOrigin, Routing.config)
+
+  // Define reconnection strategy for HTTP calls
+  private val reconnectionStrategy: ReconnectionStrategy =
+    (attempt, reason) =>
+      // Web Socket close codes: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
+      if (reason.toOption.flatMap(_.toOption.flatMap(_.code)).exists(_ === 1000))
+        none
+      else // Increase the delay to get exponential backoff with a minimum of 1s and a max of 1m
+        FiniteDuration(
+          math.min(60.0, math.pow(2, attempt.toDouble - 1)).toLong,
+          TimeUnit.SECONDS
+        ).some
+
+  // Build regular HTTP client
+  private val fetchClient: Client[IO] =
+    FetchClientBuilder[IO]
+      .withRequestTimeout(5.seconds)
+      .withCache(dom.RequestCache.`no-store`)
+      .create
+
+  // Fetch environment configuration (from environments.conf.json)
+  private val fetchConfig: IO[AppConfig] =
+    fetchClient
+      .get(ConfigFile)(_.decodeJson[List[AppConfig]])
+      .adaptError: t =>
+        new Exception("Could not retrieve configuration.", t)
+      .flatMap: confs =>
+        IO.fromOption(
+          confs
+            .find: conf =>
+              dom.window.location.host.startsWith(conf.hostName)
+            .orElse:
+              confs.find(_.hostName === "*")
+        )(orElse = new Exception("Host not found in configuration."))
+
+  // Log in from cookie and switch to staff role
+  private def enforceStaffRole(ssoClient: SSOClient[IO]): IO[Option[UserVault]] =
+    ssoClient.whoami.flatMap(userVault =>
+      userVault.map(_.user) match
+        case Some(StandardUser(_, role, other, _)) =>
+          (role +: other)
+            .collectFirst { case StandardRole.Staff(roleId) => roleId }
+            .fold(IO(userVault))(ssoClient.switchRole)
+        // .map(_.orElse(throw new Exception("User is not staff")))
+        case _                                     => IO(userVault)
+    )
+
+  // Turn a Stream[WSFrame] into Stream[ClientEvent]
+  val parseClientEvents: Pipe[IO, WSFrame, Either[Throwable, ClientEvent]] =
+    _.flatMap:
+      case WSFrame.Text(text, _) => fs2.Stream(decode[ClientEvent](text))
+      case _                     => fs2.Stream.empty
+
+  // TODO IN CRYSTAL: It would be nice to have a .useEffectWhenDepReady instead of doing .void in dependencies!
+
+  private val component =
+    ScalaFnComponent
+      .withHooks[Unit]
+      .useResourceOnMount: // Build AppContext
```

Comment:
This is a good idea, should we replicate in explore?

Reply:
We could do in explore, yes